Don't make auth file globally readable

jakewins opened the pull request:
On systems with POSIX-style file permissions,
restrict permissions on auth file to `500`.

This introduces, as a side effect, a proposed
addition to FSA to set file permissions.

jakewins committed:
Don't make auth file globally readable

On systems with POSIX-style file permissions,
restrict permissions on auth file to `500`.

This introduces, as a side effect, a proposed
addition to FSA to set file permissions.

jakewins commented:
Main driver for this is the test in `FileRepositorySerializerTest`.

tinwelint reviewed (commented):
I think it looks allright, although I'm not on a posix system. Perhaps someone else can take this for a spin and verify...

jakewins commented:
@MishaDemianenko WDYT?

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/fs/DefaultFileSystemAbstraction.java:
```diff
@@ -130,6 +134,23 @@ public boolean fileExists( File fileName )
+        Set<PosixFilePermission> posixPerms = Arrays.stream( permissions )
+                .map( p -> PosixFilePermission.valueOf( p.name() ) )
+                .collect( Collectors.toSet() );
+        Files.setPosixFilePermissions( fileName.toPath(), posixPerms );
```
as you mention in javadocs this will not gonna work on all supported operating systems

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/fs/DelegateFileSystemAbstraction.java:
```diff
@@ -115,6 +116,18 @@ public StoreChannel create( File fileName ) throws IOException
+    @Override
+    public void setPermissions( File fileName, FilePermission... permissions ) throws IOException
+    {
+        throw new UnsupportedOperationException( "This file system does not support file permissions." );
```
we should delegate to delegate file system instead since we do not know if permissions are supported there on this level

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/fs/DelegateFileSystemAbstraction.java:
```diff
@@ -115,6 +116,18 @@ public StoreChannel create( File fileName ) throws IOException
+    @Override
+    public Set<FilePermission> getPermissions( File fileName ) throws IOException
+    {
+        throw new UnsupportedOperationException( "This file system does not support file permissions." );
```
same here

MishaDemianenko commented on community/io/src/main/java/org/neo4j/io/fs/FilePermission.java:
```diff
@@ -0,0 +1,33 @@
+ */
+package org.neo4j.io.fs;
+
+public enum FilePermission
```
javadocs javadocs

MishaDemianenko commented on community/security/src/main/java/org/neo4j/server/security/auth/FileRepositorySerializer.java:
```diff
@@ -77,6 +78,15 @@ public void saveRecordsToFile( FileSystemAbstraction fileSystem, File recordsFil
+            fileSystem.create( tempFile ).close();
+            try
+            {
+                fileSystem.setPermissions( tempFile, FilePermission.OWNER_READ, FilePermission.OWNER_WRITE );
```
since this is the only usage of new functionality this can be achieved by tempFile.setRedable / tempFile.setWritable method calls.

MishaDemianenko commented on community/security/src/main/java/org/neo4j/server/security/auth/FileRepositorySerializer.java:
```diff
@@ -77,6 +78,15 @@ public void saveRecordsToFile( FileSystemAbstraction fileSystem, File recordsFil
+            {
+                fileSystem.setPermissions( tempFile, FilePermission.OWNER_READ, FilePermission.OWNER_WRITE );
+            }
+            catch(UnsupportedOperationException e)
```
not really secure

MishaDemianenko reviewed (changes requested):
It looks like desired can be achieved without introducing system dependent functionality, aslo if we introducing the possibility to set basic posix attributes we need to do the same for OS that does not support that.

jakewins commented:
@MishaDemianenko still need to retarget this for 3.5, but I changed approach a good bit and wanted your feedback. Now, instead of trying to surface POSIX-style `owner/group/all read/write/exec`, I replaced this with `AccessPolicy`, which tries to communicate intent, instead of specific flags. That way, the implementation can then set appropriate flags depending on the system, and we leave the door open to custom strategies.

I could have just surfaced setReadable/setWritable, but the intention here is that this would be a small beach-head towards better management of file permissions over all, so we can add more AccessPolicies over time.

jakewins commented:
retargeting for 3.5
